lib/base/io-engine.hpp

```diff
@@ -167,47 +167,24 @@ class AsioConditionVariable
  */
 class Timeout : public SharedObject
 {
+private:
+	boost::asio::deadline_timer m_Timer;
+
 public:
 	DECLARE_PTR_TYPEDEFS(Timeout);
 
-	template<class Executor, class TimeoutFromNow, class OnTimeout>
-	Timeout(boost::asio::io_context& io, Executor& executor, TimeoutFromNow timeoutFromNow, OnTimeout onTimeout)
-		: m_Timer(io)
+	template<class Executor, class OnTimeout>
+	Timeout(boost::asio::io_context& io, Executor& executor, const decltype(m_Timer)::duration_type& timeoutFromNow, OnTimeout onTimeout)
```

Comment:
The new implementation relies on `executor` being the current strand, doesn't it?

If so, wouldn't it make sense to additionally change the following:

1. No longer use `Executor` as a template parameter but use `boost::asio::io_context::strand` directly.
2. When the type is fixed, `strand.context()` can be used to make the `io` parameter obsolete.
3. Assert `strand.running_in_this_thread()`.